要将此笔记本转换并作为 reveal.js 演示文稿运行：```python
jupyter nbconvert SatPy.ipynb --to slides --post serve --SlidesExporter.reveal_theme=simple --SlidesExporter.reveal_scroll=True --SlidesExporter.reveal_transition=none
```

In [1]:
import warnings
warnings.filterwarnings('ignore')
import logging
logging.basicConfig(level=logging.CRITICAL)

# SatPy## 用于气象卫星图像的 Python 库
***

作者：David Hoese <sub>*发音为Haze</sub>

# SatPy 核心开发者
* 马丁·拉斯帕德* 亚当·迪布罗* 帕努·拉赫蒂宁
## 谢谢
* PyTroll 用户* 凯西·斯特拉巴拉

# 关于我
* GitHub: djhoese* 推特: djhoese* 软件开发工程师 @ 空间科学与工程中心（SSEC） - 威斯康星大学麦迪逊分校* 社区卫星处理包 (CSPP) 团队* Polar2Grid/Geo2Grid* 卫星信息熟悉工具 (SIFT) - 海报* VisPy 维护者 - 全体会议* PyTroll/SatPy 维护者

# 大纲
1. 我们使用什么？2. SatPy 如何使我们的工作更轻松？3. 使用 SatPy 可以实现的精彩功能

# 我们处理什么？气象卫星数据
* 预测和预测模型* 研究/分析* 漂亮的图片* 操作处理  
<img src="SatPy - Do Stuff Diagram.png" alt="Do Stuff Diagram">

In [1]:
from satpy import Scene
from glob import glob
scn = Scene(reader='abi_l1b', filenames=glob('/data/data/abi/20170920/*.nc'))
channels = ["C{channel:02d}".format(channel=chn) for chn in range(1, 17)]
scn.load(channels)
scn.save_datasets()
! montage C??_20170920_173040.tif -geometry 512x512+4+4  -background black montage_abi_20170920_173040.jpg

Rasterio 1.0+ required for setting colorinterp
Rasterio 1.0+ required for setting colorinterp
Rasterio 1.0+ required for setting colorinterp
Rasterio 1.0+ required for setting colorinterp
Rasterio 1.0+ required for setting colorinterp
Rasterio 1.0+ required for setting colorinterp
Rasterio 1.0+ required for setting colorinterp
Rasterio 1.0+ required for setting colorinterp
Rasterio 1.0+ required for setting colorinterp
Rasterio 1.0+ required for setting colorinterp
Rasterio 1.0+ required for setting colorinterp
Rasterio 1.0+ required for setting colorinterp
Rasterio 1.0+ required for setting colorinterp
Rasterio 1.0+ required for setting colorinterp
Rasterio 1.0+ required for setting colorinterp
Rasterio 1.0+ required for setting colorinterp


montage: Unknown field with tag 33550 (0x830e) encountered. `TIFFReadDirectory' @ warning/tiff.c/TIFFWarnings/912.
montage: Unknown field with tag 33922 (0x8482) encountered. `TIFFReadDirectory' @ warning/tiff.c/TIFFWarnings/912.
montage: Unknown field with tag 34735 (0x87af) encountered. `TIFFReadDirectory' @ warning/tiff.c/TIFFWarnings/912.
montage: Unknown field with tag 34736 (0x87b0) encountered. `TIFFReadDirectory' @ warning/tiff.c/TIFFWarnings/912.
montage: Unknown field with tag 34737 (0x87b1) encountered. `TIFFReadDirectory' @ warning/tiff.c/TIFFWarnings/912.
montage: Unknown field with tag 42112 (0xa480) encountered. `TIFFReadDirectory' @ warning/tiff.c/TIFFWarnings/912.
montage: Unknown field with tag 33550 (0x830e) encountered. `TIFFReadDirectory' @ warning/tiff.c/TIFFWarnings/912.
montage: Unknown field with tag 33922 (0x8482) encountered. `TIFFReadDirectory' @ warning/tiff.c/TIFFWarnings/912.
montage: Unknown field with tag 34735 (0x87af) encountered. `TIFFReadDirectory' 

<img src="montage_abi_20170920_173040.jpg" alt="ABI Montage" style="width:600px;">

In [15]:
from satpy import Scene
from glob import glob
scn = Scene(reader='abi_l1b', filenames=glob('/data/data/abi/20170920/*.nc'))
scn.load(['true_color'])
new_scn = scn.resample(resampler='native')
new_scn.save_datasets()
! gdal_translate -of PNG -outsize 10% 10% true_color_20170920_173040{.tif,.png}

Platform file /Users/davidh/repos/git/satpy/satpy/etc/platforms.txt not found.
Rasterio 1.0+ required for setting colorinterp
/Users/davidh/anaconda/envs/polar2grid_py36/lib/python3.6/site-packages/dask/local.py:271: RuntimeWarning: invalid value encountered in less
  return func(*args2)
/Users/davidh/anaconda/envs/polar2grid_py36/lib/python3.6/site-packages/dask/local.py:271: RuntimeWarning: invalid value encountered in greater_equal
  return func(*args2)
/Users/davidh/anaconda/envs/polar2grid_py36/lib/python3.6/site-packages/dask/local.py:271: RuntimeWarning: invalid value encountered in log10
  return func(*args2)


Input file size is 21696, 21696
0...10...20...30...40...50...60...70...80...90...100 - done.


<img src="true_color_20170920_173040.png" alt="ABI True Color" style="width:600px;">

# 我们的解决方案: SatPy
* 版本 0.9 发布* 兼容 Linux、OSX 和 Windows* 可通过 ``pip`` 和 ``conda`` (conda-forge) 安装

# SatPy 如何使我们的工作更轻松？
* 常用接口* 性能 - ``xarray`` 和 ``dask``* 社区

# 常用接口
* 阅读* 组合/修改* 重采样* 编写

# 阅读: 我如何获取我需要的信息？
* 读取卫星数据令人烦恼* 格式：NetCDF、HDF5、自定义二进制、压缩要求：
- 保留Markdown格式
- 保留代码、网址、HTML标签
- 使用标准技术术语
- 仅输出翻译，不添加解释

待翻译文本：
* 文件 "方案" 和组织* 填写值* 连接* 缺失元数据* 数据变化* 标准
<img src="https://imgs.xkcd.com/comics/standards.png" alt="drawing">

In [2]:
from satpy import available_readers
sorted(available_readers())

['abi_l1b',
 'acspo',
 'ahi_hsd',
 'amsr2_l1b',
 'avhrr_aapp_l1b',
 'avhrr_eps_l1b',
 'clavrx',
 'fci_fdhsi',
 'generic_image',
 'geocat',
 'ghrsst_osisaf',
 'grib',
 'hdf4_caliopv3',
 'hdfeos_l1b',
 'hrit_electrol',
 'hrit_goes',
 'hrit_jma',
 'hrit_msg',
 'iasi_l2',
 'li_l2',
 'maia',
 'native_msg',
 'nc_nwcsaf_msg',
 'nc_nwcsaf_pps',
 'nc_olci_l1b',
 'nc_olci_l2',
 'nc_slstr',
 'nucaps',
 'omps_edr',
 'safe_sar_c',
 'scatsat1_l2b',
 'viirs_compact',
 'viirs_l1b',
 'viirs_sdr']

# 加载数据和场景对象
* 按名称* 按波长* 按 ``DatasetID``* 指定变体* 分辨率* 极化* 校准 - 计数，辐射亮度，反射率，亮温，等

In [32]:
from satpy import Scene, DatasetID
from glob import glob

scn = Scene(reader='abi_l1b', filenames=glob('/data/data/abi/20170920/*.nc'))
scn.load(['C01', 'C02'])

In [33]:
scn.load([0.46, 0.62])

In [34]:
ds_id = DatasetID(name='C02', calibration='radiance')
scn.load([ds_id])

In [26]:
scn.available_dataset_names()

['C01', 'C02', 'C03', 'C04', 'C05', 'C06', 'C07', 'C08', 'C09', 'C10', 'C11', 'C12', 'C13', 'C14', 'C15', 'C16']

In [30]:
scn.available_dataset_ids()

[DatasetID(name='C01', wavelength=(0.45, 0.47, 0.49), resolution=1000, polarization=None, calibration='radiance', level=None, modifiers=()),
 DatasetID(name='C01', wavelength=(0.45, 0.47, 0.49), resolution=1000, polarization=None, calibration='reflectance', level=None, modifiers=()),
 DatasetID(name='C02', wavelength=(0.59, 0.64, 0.69), resolution=500, polarization=None, calibration='radiance', level=None, modifiers=()),
 DatasetID(name='C02', wavelength=(0.59, 0.64, 0.69), resolution=500, polarization=None, calibration='reflectance', level=None, modifiers=()),
 DatasetID(name='C03', wavelength=(0.8455, 0.865, 0.8845), resolution=1000, polarization=None, calibration='radiance', level=None, modifiers=()),
 DatasetID(name='C03', wavelength=(0.8455, 0.865, 0.8845), resolution=1000, polarization=None, calibration='reflectance', level=None, modifiers=()),
 DatasetID(name='C04', wavelength=(1.3705, 1.378, 1.3855), resolution=2000, polarization=None, calibration='radiance', level=None, modifi

In [29]:
scn['C02']

<xarray.DataArray (y: 21696, x: 21696)>
dask.array<shape=(21696, 21696), dtype=float64, chunksize=(4096, 4096)>
Coordinates:
  * x        (x) float64 -0.1519 -0.1519 -0.1518 -0.1518 -0.1518 -0.1518 ...
    y_image  float32 0.0
    x_image  float32 0.0
  * y        (y) float64 0.1519 0.1519 0.1518 0.1518 0.1518 0.1518 0.1518 ...
Attributes:
    satellite_longitude:    -89.5
    satellite_latitude:     0.0
    satellite_altitude:     35786.0234375
    units:                  %
    modifiers:              ()
    wavelength:             (0.59, 0.64, 0.69)
    grid_mapping:           goes_imager_projection
    _Unsigned:              true
    cell_methods:           t: point area: point
    standard_name:          toa_bidirectional_reflectance
    scale_factor:           0.158592
    ancillary_variables:    []
    valid_range:            [   0 4094]
    long_name:              ABI L1b Radiances
    add_offset:             -20.2899
    resolution:             500
    calibration:            

In [14]:
scn.show('C01')

In [4]:
scn.save_dataset('C01', base_dir='/Users/davidh/repos/git/pytroll-examples/satpy/presentations/2018_scipy/', writer='simple_image')

! gdal_translate -of PNG -outsize 10% 10% C01_20170920_173040.png C01_20170920_173040.small.png

Input file size is 10848, 10848
0...10...20...30...40...50...60...70...80...90...100 - done.


<img src="C01_20170920_173040.small.png" alt="ABI C01" style="width:550px;">

# 组合：如何制作漂亮的彩色图片？
<img src="abi_rgbs_20170920_173040.jpg" alt="ABI RGB montage" style="width:550px;">

In [ ]:
from satpy import Scene
from glob import glob

scn = Scene(reader='abi_l1b', filenames=glob('/data/data/abi/20170920/*.nc'))
scn.load(['overview', 'natural', 'airmass'])
new_scn = scn.resample(resampler='native')
new_scn.save_datasets()
! montage {true_color,overview,natural,airmass}_20170920_173040.tif -geometry 512x512+2+2 -background black abi_rgbs_20170920_173040.jpg

Platform file /Users/davidh/repos/git/satpy/satpy/etc/platforms.txt not found.
Rasterio 1.0+ required for setting colorinterp
Rasterio 1.0+ required for setting colorinterp
Rasterio 1.0+ required for setting colorinterp
/Users/davidh/anaconda/envs/polar2grid_py36/lib/python3.6/site-packages/dask/local.py:271: RuntimeWarning: invalid value encountered in less
  return func(*args2)
/Users/davidh/anaconda/envs/polar2grid_py36/lib/python3.6/site-packages/dask/local.py:271: RuntimeWarning: invalid value encountered in greater_equal
  return func(*args2)


In [1]:
from satpy import Scene
from glob import glob

scn = Scene(reader='abi_l1b', filenames=glob('/data/data/abi/20170920/*.nc'))
scn.available_composite_names()

['airmass', 'ash', 'dust', 'fog', 'green', 'green_raw', 'green_snow', 'ir_cloud_day', 'natural', 'natural_raw', 'natural_sun', 'night_microphysics', 'overview', 'overview_raw', 'true_color', 'true_color_raw']

In [2]:
scn.load(['airmass'])
scn['airmass']

<xarray.DataArray 'concatenate-d3ffb434ba4f6164ba94e209a0860c74' (bands: 3, y: 5424, x: 5424)>
dask.array<shape=(3, 5424, 5424), dtype=float64, chunksize=(1, 4096, 4096)>
Coordinates:
  * x        (x) float64 -0.1518 -0.1518 -0.1517 -0.1517 -0.1516 -0.1516 ...
    y_image  float32 0.0
    x_image  float32 0.0
  * y        (y) float64 0.1518 0.1518 0.1517 0.1517 0.1516 0.1516 0.1515 ...
  * bands    (bands) <U1 'R' 'G' 'B'
Attributes:
    long_name:               ABI L1b Radiances
    grid_mapping:            goes_imager_projection
    level:                   None
    satellite_latitude:      0.0
    sensor:                  abi
    satellite_altitude:      35786.0234375
    ancillary_variables:     []
    start_time:              2017-09-20 17:30:40.800000
    cell_methods:            t: point area: point
    polarization:            None
    resolution:              None
    area:                    Area ID: some_area_name\nDescription: On-the-fly...
    platform_name:           GOES

# 组合 YAML 配置文件
```yaml
  airmass:
    compositor: !!python/name:satpy.composites.Airmass
    prerequisites:
    - 6.2
    - 7.3
    - 9.7
    - 10.3
    standard_name: airmass
```

# 修饰符：如何让我的漂亮图片更漂亮？
<img src="true_color_before_after_20170920_173040.jpg" alt="ABI True Color Before/After">

# 完成不可能的任务

In [ ]:
scn.load(['true_color'])

涉及的操作：
* 3 个输入波段（500米 和 1000米）* 3个复杂数据增强* 1 伪波段 (绿色)* 上采样到500米* 最终结果：21696 行 \* 21696 列 \* 8 字节 \* 3 输入波段 = ~10.5 GB* 基于Dask的SatPy在6至8分钟内加载、计算并保存为GeoTIFF格式

In [1]:
from satpy import Scene
from glob import glob

scn = Scene(reader='abi_l1b', filenames=glob('/data/data/abi/20170920/*.nc'))
scn.load(['true_color_raw'])
new_scn = scn.resample(resampler='native')
new_scn.save_datasets()
! montage true_color_raw_20170920_173040.tif true_color_20170920_173040.tif -geometry 512x512+2 -background black true_color_before_after_20170920_173040.jpg

# 修饰符
* 加载期间
```python
ds_id = DatasetID(name='C01', modifiers=('sunz_corrected',))
scn.load([ds_id])
```

* 配置复合材料
```yaml

  true_color:
    compositor: !!python/name:satpy.composites.SelfSharpenedRGB
    prerequisites:
    - name: C02
      modifiers: [sunz_corrected, rayleigh_corrected]
    - name: green
    - name: C01
      modifiers: [sunz_corrected, rayleigh_corrected]
    standard_name: true_color

```

# 重采样
<img src="remap.png" alt="Remapping">

<h6><a href="https://openi.nlm.nih.gov/detailedresult.php?img=PMC4299089_sensors-14-23822f3&req=4">参考文献</a></h6>

* 用户希望有一个与极地轨道无关的通用投影* 重采样以重新投影到一个通用投影

In [36]:
from satpy import Scene
from glob import glob
scn = Scene(reader='viirs_sdr',
            filenames=glob('/data/data/viirs/conus_day/*t1801*.h5'))
scn.load(['I04'])

In [ ]:
scn.show('I04')

<img src="I04_20120225_180125.swath.png" alt="VIIRS I04 Swath">

In [ ]:
scn = Scene(reader='viirs_sdr',
            filenames=glob('/data/data/viirs/conus_day/*t180*.h5'))
scn.load(['I04'])
new_scn = scn.resample('211e')
new_scn.show('I04')

<img src="I04_20120225_180125.png" alt="NPP VIIRS I04" style="width:450px;">

In [ ]:
from pyresample.geometry import AreaDefinition
area_def = AreaDefinition(
    'name', 'description', 'merc',
    {'proj': 'merc', 'datum': 'WGS84', 'lon_0': -95.0},
    1000, 1000, (0, 1750000, 4500000, 5500000))

new_scn = scn.resample(area_def, radius_of_influence=10000)
new_scn.show('I04')

<img src="I04_20120225_180125.custom_area.png" alt="NPP VIIRS I04 Custom" style="width:450px;">

# 保存到磁盘
* 作者* 地理TIFF (rasterio)* Pillow (png, jpeg 等)* AWIPS SCMI* 符合CF规范的NetCDF（有限的投影支持）* 尽可能使用 dask 块

In [63]:
from satpy import available_writers
available_writers()

['mitiff', 'simple_image', 'geotiff', 'cf', 'scmi']

In [ ]:
scn.save_datasets(writer='geotiff', base_dir='/tmp', compress='LZW')

# 基础知识
```python
from satpy import Scene
scn = Scene(reader='abi_l1b', filenames=[...])
scn.load(['C01', 'C02', 'C08', 'true_color', 'air_mass', 'natural'])
new_scn = scn.resample(resampler='native')
new_scn.save_datasets(base_dir='/tmp')
```

# 趣味内容：堆叠场景

In [2]:
from satpy import Scene, MultiScene
from glob import glob
import os
from dask.diagnostics import ProgressBar
base_dir = '/data/data/viirs/2018_06_29/'
viirs_passes = ['2018_06_29_180_0614', '2018_06_29_180_0752', '2018_06_29_180_0934']
viirs_scenes = [Scene(reader='viirs_sdr',
                      filenames=glob(os.path.join(base_dir, vp, '*.h5'))) 
                for vp in viirs_passes]

mscn = MultiScene(viirs_scenes)
mscn.load(['I04'])
new_mscn = mscn.resample('211e')
blended_scn = new_mscn.blend()
with ProgressBar():
    blended_scn.save_datasets()

[########################################] | 100% Completed |  4min 15.4s
[########################################] | 100% Completed |  0.1s


In [4]:
! gdal_translate -of PNG -outsize 10% 10% I04_20180629_061415{.tif,.png}

Input file size is 5120, 5120
0...10...20...30...40...50...60...70...80...90...100 - done.


<img src="I04_20180629_061415.png" alt="Blended I04" style="width:550px">

# 有趣的内容：动画
```python
scenes = [Scene(reader='abi_l1b', filenames=step_files)
          for step_files in grouped_files]
mscn = MultiScene(scenes)
mscn.load(['C01'])
new_mscn = mscn.resample(resampler='native')
new_mscn.save_animation('{name}_{start_time:%Y%m%d_%H%M%S}.mp4',
                        fps=5, batch_size=4,
                        output_params=["-vf", "format=yuv420p,scale=640:-1"])
```

In [5]:
from IPython.display import HTML
HTML("""
<video width="640" height="480" controls>
  <source src="C01_20180623_000045.mp4" type="video/mp4">
</video>
""")

# 趣味内容: Cartopy

In [ ]:
from satpy import Scene
from glob import glob
import matplotlib.pyplot as plt

filenames = glob('/data/data/viirs/conus_day/*t180*.h5')
scn = Scene(reader='viirs_sdr', filenames=filenames)
scn.load(['I04'])
new_scn = scn.resample('211e')

crs = new_scn['I04'].attrs['area'].to_cartopy_crs()
ax = plt.axes(projection=crs)

ax.coastlines()
ax.gridlines()
ax.set_global()
plt.imshow(new_scn['I04'], transform=crs, extent=crs.bounds, origin='upper')
cbar = plt.colorbar()
cbar.set_label("Kelvin")
plt.show()

<img src="I04_cartopy_plot.png" alt="I04 Cartopy">

# SatPy 与 PyTroll 社区
* 社区网站: http://pytroll.github.io/* GitHub：https://github.com/pytroll/satpy* SatPy 文档: http://satpy.readthedocs.io/en/latest/* 邮件列表: https://groups.google.com/d/forum/pytroll* 一年两次研讨会！* 在 Slack 上与我们聊天 (社区网站上的邀请链接)
<img src="https://raw.githubusercontent.com/pytroll/pytroll/main/web/source/images/pytroll_dark_small.png" style="background-color:black;">


In [5]:
from IPython.display import IFrame
IFrame('https://www.youtube.com/embed/eBQi2G_fqXQ?rel=0', width=600, height=400)